### DISCUSSION TITLE
Monorepos

### DISCUSSION THREAD

**Discussion by Shadester** [MobileNativeFoundation/discussions#31] [General]
Since our organization is starting to grow, backend and frontend has started using a monorepo. We have been asked to move the mobile code (both iOS and Android) into it. I have been against it, due to the extra tools needed to handle it, the codebases are in non compatible programming languages and since we are not doing releases on a specific cadence.

I am curious to hear what other companies are doing. Are you using monorepo for the entire organization or monorepo for only mobile or no monorepos? And what are your rationale for choosing one system over another?

**Comment by BalestraPatrick** [comment]
This is a very interesting one. Whether or not you should have a monorepo depends on many factors. I can try to describe why we're choosing to move all our client code to a monorepo.

In our organization, all the iOS code and apps have been historically in different repos. All the iOS apps had their own repo. All the build tooling was getting duplicated for each new repo, and developers would end up changing things in each repo diverging even more from what the tools were initially built for. For example, we started developing an SDK that was supposed to be used by every single app. This SDK couldn't live in our main repo since it needed to be depended upon by other apps living in different repos too. So the team decided to create a separate repo, copy over some build tooling and use submodules to integrate their source code with other repos. This is very painful in the long run because it leads to a so called "bump-hell". Make a change, bump it into the consumer, and realize it doesn't work as it should. Go back to the original repo, make one more change, and bump it again. This is very frustrating for developers who just want to iterate fast.
Of course there are some advantages in working in isolation in different repos. Usually the team is much more in control of what they're doing and their PR cycle would probably be faster (if you don't have some kind of mechansim to dynamically build and test what changed in pre-merge).

**We asked our developers what they would prefer in a survey when confronted with the decision to go with a monorepo or not**:
- **Push vs Pull**: do you want to push changes to your consumer or have them pull?
- **Consistency vs Independence**: do you prefer consistency across all our projects by trading off a bit of speed at first (with maybe more speed in the future due to strict alignment), or do you prefer to be independent and move faster in the short-run?
- **Internal vs Global Verification**: do you care about global verification? In other words, do you want to verify your changes against all possible consumers?
- **Internal vs Global Changes**: should we optimize for low friction local changes but harder global changes or the reverse?

These questions helped us decide if we should go for a monorepo or instead a poly-repo layout. 

In the end, we decided to go for a monorepo layout and this is what we're currently building. We also feel that code reuse is higher both inside the same platform, as well as across platforms. For example, linting and formatting could be unified in a single tool for all the code in a monorepo. Even though there are platform specific tools, there's always a base infrastructure layer used to run these tools which can be shared. Moreover, using a single build system can decrease even more platform-specific tooling and increase code reuse. Using a build system like Bazel allows you to make it super easy for developers not familiar with the 5 different languages in your monorepo how to build them, which might be much harder in a distributed repo layout. 

There are so many different ways to build a monorepo too. You may still want to move all the code inside the same repo, but keep tooling distinct initially. When the opportunities arise, you may then benefit for free from tools that other parts of the organization have built already for you, without having to spend months adapting tools that were built just for one specific use-case.

**Reply by Shadester** [reply]
Thanks for the answer! Many good points. But it sounds like you are using a mobile monorepo? Or maybe even just an iOS monorepo? Not for backend/web?

In our organization they want to dump all the code (mobile, backend, web, API, embedded, etc) into one repo.

**Reply by BalestraPatrick** [reply]
For now it's a mobile monorepo (iOS, Android, C++), but it could be extended to other disciplines in the future. You may need substantial research and investment to maintain a huge repo including all the disciplines based on their size, that's something to keep in mind.

**Reply by julioz** [reply]
I'm curious on how you manage the social side of that mobile monorepo - do you have specific rules for assigning PR reviews, ownership over sub-system/framework/module; etc? In other words, in such a large repository, how can you quickly know that team X (or person Y) is responsible for a build speed regression, test failure or infrastructure instability?

**Reply by lukewar** [reply]
It's a great discussion! 

> In other words, in such a large repository, how can you quickly know that team X (or person Y) is responsible for a build speed regression, test failure or infrastructure instability?

@julioz in case your team is using GitHub [Code owners](https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners) feature could help with addressing some of those challenges.

**Reply by hendych** [reply]
Hi @BalestraPatrick , you mentioned about the cons of bump hell. What if we use separate repo and git submodule in repos that would depend on the SDK?

**Comment by shepting** [comment]
We have a monorepo for backend services, another one for frontend, and then one for our mobile apps. We've really appreciated being able to shared tooling across iOS/Android even if we are not sharing code itself. Things like flaky test detection, etc are all similar across platforms so using a shared repository makes a lot of sense.

I'm quite in favor of monorepos.

**Reply by rudro** [reply]
We also have a multi-monorepo setup, unfortunately iOS and Android are not in the same repo. There are several bits (not code) that are shared across the two mobile platforms and as Steven said, tooling that would really benefit from combining the two repos.

A (somewhat old) post about the iOS monorepo move : https://eng.uber.com/ios-monorepo/

Also in favor of monorepos.

**Reply by Shadester** [reply]
I think multi-monorepos makes more sense than one for all. So mobile can share one, for example. Having everything in one seems much harder.

**Reply by keith** [reply]
We're also the same boat of wanting to share tooling across iOS and Android but having separate repos for them. I have yet to find the compelling reason to disrupt developers to merge them though. I'm also potentially worried about how git handles that long term since even just `git status` can take ~2s in our Android repo. What has been your experience there?

**Reply by erikkerber** [reply]
@Keith Have you ever cleaned out Git history on any of your repos (through git-filter or similar)?

A year or so back our clones took about ~45 minutes, and it was far more than the network serving as the bottleneck, with the skeletons of many past binaries and large images making packing and other Git operations painfully slow.

Over one slow holiday week when most were OOO, we executed a well-coordinated and communicated run of [Git BFG](https://rtyley.github.io/bfg-repo-cleaner/), zapping old committed libs and 2MB+ images from history. Our clones went down to ~10 seconds, and `git reflog` no longer took an entire night. I wonder if a similar cleanup would make `git status`, and maybe some other operations play nicely. Educating an entire mobile army on what a total history rewrite means to them is, of course, a delicate dance.

Of course, ditto to any monorepo that might accumulate detritus over many years and many commits.

**Reply by keith** [reply]
We haven't done anything like this yet. We do have some tooling to make sure folks don't check in files >1mb (at least accidentally, it's allowed if you really need to), but we might still do this in the future.

**Reply by dalemyers** [reply]
Presumably if your limit is 1MB, your pbxproj is generated? I think ours is about 5MB at this point.

**Reply by keith** [reply]
Yes ours is. We also use "focused" projects where they only contain what folks ask for, so they're still pretty small.

**Reply by jgavris** [reply]
@keith Check out my `git core.fsmonitor` plugin, it makes `git status` really fast. Fast enough that I have it in my terminal prompt in a monorepo. https://github.com/jgavris/rs-git-fsmonitor

> We're also the same boat of wanting to share tooling across iOS and Android but having separate repos for them. I have yet to find the compelling reason to disrupt developers to merge them though. I'm also potentially worried about how git handles that long term since even just `git status` can take ~2s in our Android repo. What has been your experience there?

**Reply by keith** [reply]
Thanks for the link, I have tried this but haven't actually seen any speed up with it

**Reply by dalemyers** [reply]
I've also investigate this (as well as using the Perl one that Git provides), but also saw no difference at all. A brief investigation showed that watchman was returning the paths practically instantly, but then git was appearing to just ignore that and do its own thing anyway. @jgavris do you know what is going on here?  (I know this isn't necessarily the place for technical discussions like this, so feel free to email me using my first name @ my last name + ".io" if you prefer.)

**Reply by jgavris** [reply]
@keith How many files are in the repo you're testing on? Is the repo on a network or otherwise unsupported by watchman filesystem by any chance? Enabling the monitor on my repo cuts the stat calls in half roughly. You might have to wait a second or run `git status` a few times for `watchman` to update its index / clock to get the 🚀 . 

```shell
jgavris in ~/code/software on develop
λ sudo dtruss git status 2>&1 | grep stat | wc -l

    1101
jgavris in ~/code/software on develop
λ git config --unset core.fsmonitor
jgavris in ~/code/software on develop
λ sudo dtruss git status 2>&1 | grep stat | wc -l

    2554
```

**Reply by keith** [reply]
looks like it does go from ~2500 -> ~100 with it on, but still not seeing a performance difference

**Comment by dalemyers** [comment]
There's a couple of responses here where people have gone for a monorepo. I can give our experiences going the other way. I work on Outlook for iOS, and we have individual repos for each app (iOS and Android), then for any shared tool (such as our rendering engine, etc.). 

Patrick has some great points above about "bump hell" and being able to share tooling. Those are definitely the massive advantages of having a monorepo. For contrast, we have to setup the same tooling suite for every repo, and that can take time. Updating a dependency only goes to a couple of repos at most, so that ones not too bad. 

What are the advantages though? 

Well, by having things separate, we've found that it's much easier to enforce clean separation between components. You can't accidentally couple things when they are in separate repos (tooling is a solution to this, but requires a greater effort at the outset to enforce). Not being able to rely on building the app and using the apps test suite when you push a change forces you to write comprehensive tests for components, and then do the integration tests on the apps. 

We've found that our performance is much higher. When build, we only have to clone what we need, not a whole bunch of extra code for other apps. And since our repo is smaller, we aren't going to hit the same git performance issues (that said, we are at about 2s anyway for a `git status` after updating to Big Sur - something we are actively working on reducing). And then of course, because we are smaller, we only have to build that app/component/whatever, rather than having to "build the world" (although, again, this can be avoided through smarter tooling). 

I've worked on monorepos and individual repos. They both have their pros and cons. There is no one size fits all solution. The questions that Patrick asks to decide are spot on I think with helping you make your decision.

**Comment by adamhowardprice** [comment]
At [Butterfly Network](https://www.butterflynetwork.com), we have a monorepo containing the code for the iOS app, Android app, shared C++ code, firmware, data warehouse, manufacturing test code, and pretty much everything else other than web/backend.

There is a lot of useful tooling to make this work for us.
- Git LFS for all large files and binaries. Yes the repo is big but you only have to clone it once and you don't need to pull most things down from LFS. A local [.lfsconfig](https://github.com/git-lfs/git-lfs/blob/main/docs/man/git-lfs-config.5.ronn) file is handy.
- We've written a tool called DynamicBuildkite (in Swift!) to allow us to build and test only the minimum required set of CI jobs required based on a PR. This means if you're only changing a leaf module, you can get CI to pass in a few seconds, whereas postmerge builds that build everything take about a half hour. The tools takes a YAML config file that describes the dependency graph and a customized Buildkite pipeline YAML file. Since this is the entry point for our CI, we use it for a number of other handy monorepo tasks, such as opting in to specific CI jobs using a special commit message or PR title, or running special CI pipelines via environment variables.
- Shared bootstrapping infrastructure allowing you to set up only the subset of the repo that you are going to be working in.
- Shared Ansible code for provisioning machines, downloading developer tools, setting up Python virtualenvs, configuring remote machines, etc.
- Shared CI jobs for common tasks like labeling PRs, shared static analysis, etc.

All of the trade-offs @BalestraPatrick points out are valid. I know this because I was weighing the pros and cons on this subject when I was still at Spotify (pre-monorepo) and those are the ones I came up with. Now that I'm on the other side, the choice seems much clearer.

"Push" is better than "Pull". If you have ten teams that require pulling their changes in one at a time, you are constantly incurring risk. By removing the expectation that all ten modules work together at HEAD, allowing teams to work more independently, you have to constantly reverify that they work together by running integration tests on bumps. Logical conflicts are common in this multi-repo scenario, meaning you spend a lot more time debugging these conflicts every time you bump. These conflicts are annoying if you bump a lot, and if you bump seldom, the conflicts get bigger as you spend more time unintegrated. All this bumping wastes a ton of time for both feature teams as well as infra teams.

As an infra team, it's hard to overstate the advantage of sharing all infrastructure in the same repo as opposed to needing to worry about duplicating efforts across repos. In all the time we don't spend thinking about differences in tooling, submodule bumps, diamond dependency problems, branch hell, integration testing separate repos, we can spend that extra time building cool useful tools to help people.

As our dev team scales up, it's clear that consistency is a huge benefit that is only going to grow. There are other ways for teams to exercise independence in a monorepo... and they do. Feature flags, branches, etc. I don't hear many people complaining about wanting more independence or wanting to stay ahead of the curve. On the other hand, it does mean that we as an infra team have a high degree of responsibility to make sure the repo is up-to-date with the latest versions of basic tools like Xcode, Android Studio, etc.

Code separation via modularization is a separate issue from repo structure. You don't need a separate repo to architect your code in a modular way. That said, being in the same repo does make it easier to break the modular architecture. We might, in the future, choose to write a tool to catch imports that cross module boundaries.

As Patrick points out, making global changes is more difficult in a monorepo. We've been working on a branch for a while now to upgrade our version of Python, and it's required some work to keep this branch up-to-date by merging changes into it over time. It's also going to be quite a chore reviewing that big change. However, once it's done, the benefits will be immediate for everyone, and most people didn't have to worry about it at all.

Another gap is the ability to release subcomponents of a monorepo in open source. I believe there is the potential to use a tool like Google's [Copybara](https://github.com/google/copybara) to sync a separate repo from a directory, but we haven't found the time to focus on this yet. If anyone here has experience open-sourcing subcomponents of their monorepo using that tool or another approach, I'd love to hear about it. I'm tempted to open-source DynamicBuildkite if other people are interested, since it's relatively uncoupled from our internal code.

**Reply by raven** [reply]
I'm super keen to hear more about DynamicBuildkite

> We've written a tool called DynamicBuildkite (in Swift!) to allow us to build and test only the minimum required set of CI jobs required based on a PR. 

Have you written up / presented / open sourced this at all?

**Reply by keith** [reply]
We have some tool like this internally as well, we use ours with bazel today but also used it without it in the past. The tl;dr of ours is:

1. hit the github API to determine what files were changed
2. Use bazel queries to determine what apps / tests could have changed given those
3. Generate a list of buildkite jobs to run based on this using some sharding logic around splitting jobs 1 per app, 1 test job with the app specific dependencies etc (you could definitely do something smarter here, but this is some signal for weighting the build graph based on what apps depend on them with the assumption that the targets will heavily overlap in their dependency trees)
4. upload the generated pipeline to buildkite

**Reply by tinder-maxwellelliott** [reply]
> I'm super keen to hear more about DynamicBuildkite
> 
> > We've written a tool called DynamicBuildkite (in Swift!) to allow us to build and test only the minimum required set of CI jobs required based on a PR.
> 
> Have you written up / presented / open sourced this at all?

There is a bazel approach for this open sourced here https://github.com/Tinder/bazel-diff

**Comment by franciscoamado** [comment]
👋🏻 Hope I'm not late to the party. A little touch on the Monorepo experience at [Farfetch](Farfetch.com)

Right from the beginning, [Farfetch](Farfetch.com) iOS was built with a modular strategy. We build most of our API abstractions, Utility extensions, Design tokens and such independently, on its own repositories. It allows us to share this with multiple iOS Apps whilst forcing its contributions to be decoupled from any Apps' specific logic. 

Specifically, in our main Shopping App, we've been through all the paths. 
We've tried a Single project in one repo but this wouldn't scale as we needed at the time. Corners were cut which led to not-so-good practices.

Then we moved to a multi-project multi-repo approach. The separation of concerns between areas helped us implement better patterns but the multi-repo approach was more cumbersome than what we expected. The information got scattered into multiple Pull Requests making the review of a feature even more difficult. Useful information was pushed out to the edges and when adopting the code in our main Project we've ended up only seeing bumps in dependencies, causing the Bump Hell a few mentioned. Our CI tooling was also replicated everywhere making it difficult to keep it aligned. 
Maybe a too strict of a solution for what we wanted.

Finally, we've landed into a multi-project monorepo approach, by centralizing our most "prone to change" code.  We maintain the same modularization techniques but in a monorepo style. 
With more atomic Pull Requests, our code reviews got more contextualized and time to merge got reduced significantly. Our productivity and satisfaction increased. Tooling is now much more consistent, easier to change and adopt. For the frameworks included in the monorepo, no bumps are required but we still get the separation of concerns since we're respecting our framework interfaces.
Interestingly, we've chosen to maintain our CI much like it was with separated repositories. When a PR is opened, our CI infrastructure is configured to only test the respectable projects that are being modified and our App's project for validating its integration. (Much like Keith mentioned for Lyft, but at a smaller scale and without Bazel).
Git hasn't been a problem at this point in our scale but applying shallow clones in CI has helped a lot with the fetch time.

But we keep looking at the future and thinking how we could integrate ideas like module caching, automatic dependency resolution, project generation and possibly another build system 😉  to help us get to the next level.

**Comment by FranzBusch** [comment]
Hey 👋

Also want to share our experience at Sixt for our iOS applications and how we evolved over the past 4 years. Back in the days, we had three Objective-C based apps that only shared code between them by copy-pasting. At that point, we decided that we are going to re-write our apps using Swift and wanted to share code between them. We started by creating a repository for each app and for each library. This worked out pretty nicely in the beginning and we had clear ownership and separation of concerns for our individual modules.

However, after some time we began to feel multiple pain points with this multi-repo setup.
- Opening PRs to multiple repos before you can work on a story and getting reviews for all of them slowed us down
- Transitive dependencies across libraries made this even worse
- When refactoring a lower level module it was hard to see all the use-sites right away

We then decided to move to a mono-repo that includes all internal iOS libraries and apps and since then haven't looked back. The key point for us was finding a good folder structure, clear GH code ownership, and using XcodeGen to remove `.xcodeproj` merge conflicts.

**Comment by maciejpiotrowski89** [comment]
Hi 👋,

I noticed that some of the folks here might be interested in the topic of **shrinking size of a Git monorepo**. We did it in [Allegro](allegro.pl) - an e-commerce platform and the most popular product search-engine in Poland. We have an iOS app and its code was **migrated to a monorepo a few years back** - before that we used to have a polyrepo for its source code. When it comes to the **cons of a polyrepo** the following, that were already mentioned in this [discussion](https://github.com/MobileNativeFoundation/discussions/discussions/31), were the most crucial for us:
- problems with management of dependencies scattered throughout multiple repos
- difficulties to verify that interface contract between modules is maintained upon dependency version update
- inability to unify the tooling

When we migrated the code to a monorepo we were able to address a different set of issues, among which were slow compilation times of the source code. We were able to use Bazel 💚 as a part of our unified tooling - I explain how we us it in the talk [Speeding up the build process of a monolithic application](https://www.youtube.com/watch?v=Jx4jfdle4KQ).

# Shrinking the size of a monorepo

Our main repository for the iOS application have become our monorepo. After 9 years of development of the app the repo size have grown enormously and the `git clone` command became a nightmare taking too much time. We had a possibility to shrink its size during the migration from an on-premise to an external git repo hosting provider. 
  
### General repo scale 

The  General scale of our old repository was as follows:

- almost 9 years of history with  **91k +** commits
- **440k** [BLOBs](https://en.wikipedia.org/wiki/Binary_large_object) were stored in the repo: multiple `.png` and `.jpeg` files, 3rd-party frameworks and toolset binaries. The unpacked size of the BLOBs would add up to **36 GB** and the biggest BLOB stored was **100+ MB**
- **680k** git  *tree* objects
- the unpacked repo size on the main development branch was **8+ GB**  where the `.git` dir size after a clone was  **7+ GB**. The `.git` directory contained compressed `.png` (**4.5+ GB** 🤯) and `.pbxproj` (**600+ MB**) files

### New repo scale

After the migration and history rewrite we shrink the repo size to:

- **71k +** commits
- **230k** [BLOBs](https://en.wikipedia.org/wiki/Binary_large_object), where all BLOBs unpacked would add up to **1.6 GB** - this number includes also size of source code files, whereas assets and binaries were migrated to an external storage
- **455k+** git  *tree* objects

## How did we do it?

The history-rewrite process required proper planning and a few steps:
1. Analysis of the old repo contents and its history
1. Creating a reproducible procedure for the history rewrite
1. Dry running the procedure to test out the process
1. Planning and scheduling activities involved to migrate the repo 
1. Proper communication about the process to stakeholders (a.k.a. developers 👨‍💻)
1. The actual migration

### The repo analysis

Goals:
- find items that can be removed from the history 
- select items that can be migrated to an external storage

We used tools such as [git-sizer](https://github.com/github/git-sizer) and [git-filter-repo tool](https://github.com/newren/git-filter-repo) to get information about types of files stored in the repository. If you wanted to do the same the workshops from [GitHub Universe](https://githubuniverse.com/professional-services-workshop-2-how-to-keep-git-monorepos-manageable/) and [the scripts introduced there](https://github.com/githubuniverseworkshops/grafting-monorepos/issues/2) might be a good starting point.
 
 From the analysis we were able to select the following items for entire removal from the history:

- deleted dirs and files
- unwanted paths: e.g. `Pods/`, invalid symlinks causing deep nesting of paths
- unwanted history of paths: e.g.  `Vendor` for storing 3rd party dependencies or  `Toolset` with binaries
- unwanted files: .e.g `.pbxproj`that can be generated by [XcodeGen](https://github.com/yonaskolb/XcodeGen) and their history is meaningless (**600+ MB** savings in our case), history of BLOB files  such as `.jpg`, `.png`, `.a`, `.dylib`, `.pdf`, `.zip`, `.mp4`, `.json` 

We decided to track BLOBs by the [Git LFS \(Large File Storage\)](https://git-lfs.github.com/). In our case the following were a good suit for it:
- large binary files  `.jpg`, `.png`, `.a`, `.dylib`, `.pdf`, `.zip`, `.mp4`, `.json` 
- framework binaries 
- toolset binaries 

### Reproducible procedure and dry runs

We created a script that contained all commands that removed redundant items from the history. To remove items we used [git filter-repo](https://github.com/newren/git-filter-repo) - it's way much more performant than git's built-in `git filter-branch` (do not use it!). Some examples of usage: 
```
git filter-repo --invert-paths --path Pods/ --force 
git filter-repo --invert-paths --paths-from-file remove.txt --force
git filter-repo --invert-paths --path-glob '*.pbxproj' --force
```
After the removal we restored the most-recent version of binaries, frameworks and BLOBs to the repo and we tracked them with Git LFS:
```
git lfs track "*.png"
git lfs track "Vendor/SomeSDK/SomeSDK.framework/SomeSDK"
```

We have run the script a few times to verify the output size of the repo. One crucial aspect after the run was to verify that **all plans** on the CI (Continuous Integration) pass - we did it to check that the app still compiles, tests pass and no more files that the ones we had wanted were actually deleted. 

### Communication 

Crucial aspect of introducing any change is communication. It's good to prepare it in advance, have team members to review it. We used a few channels so that our devs would get important info about the migration and history rewrite on the channel that suits best their working habits (e-mails, instant messaging tool, dev forums).

## Some final thoughts

It took a lump time to prepare the migration, understand the history of the repository and select proper items and strategies for the migration. The links here might be a good starting point if you wanted to rewrite histories of your over-weighted repos:

- [GitHub Universe Workshops](https://githubuniverse.com/professional-services-workshop-2-how-to-keep-git-monorepos-manageable/)
- [git-sizer](https://github.com/github/git-sizer) 
- [git filter-repo](https://github.com/newren/git-filter-repo)

When creating a plan for the rewrite remember to have a checklist that you can use to verify outcomes and to remember all the steps involved in the process. Plan the history rewrite (along with the migration, which was in our case) for the time that folks would not want to push code to the repo. We used Friday evening, and yes, we had to fix some issues over the weekend - not everything went smoothly.

You can have on the servers a copy of your old repository in a READ ONLY mode - it will serve as a **backup** and will contain the actual history.

**Reply by maciejpiotrowski89** [reply]
@erikkerber you asked about cleaning out git history in your [comment](https://github.com/MobileNativeFoundation/discussions/discussions/31#discussioncomment-434363). The post above might be interesting for you 🙂 .

**Reply by vfn** [reply]
@maciejpiotrowski89 thanks for sharing your experience. I went through a similar process some time ago, never got to implement it though. We went through the planing phase and tried a few approaches, including rewriting history to migrate files to Git LFS, but at the end the proposed solution was never implemented. But I have to say that on that case the repo was not as massive as this, but it was growing fast and could eventually become a more painful problem.

It might be a controversial thought, but have you considered migrating to a new repo completely. What I mean is, in an ideal scenario, you'd simply make the current repo read-only. On your local machine, delete all git history, and start fresh with initial commit containing just the code that was at the top of the main branch of the _legacy_ repo. Push this new history to the new read-write remote, share that new remote link with any _stakeholder_ and ask them to migrate or start new branches from this new remote's main branch.

I have never done that but I'd be curious if anyone has taken that approach or at least investigated it and can share some pros/cons.

Yes, on a normal scenario the above would require more work, as in migrating working branches, make sure big files are moved to Git LFS if applicable, and many other things that could complicate the process.

The benefits on my view are:
- you could every year, or every _x_ years, do that and your repo size could/would decrease by a lot.
- anyone interested in the whole history can add the many "legacy" remotes to their local copy and they would be able to see the "whole" history of the project, as I believe most Git GUI, when sorting chronologically, would show a full representation of all the remotes' histories. Most people wouldn't be interested in going that far back in the project history, unless you're auditing something.

Of course this is a simplified view of the problem/solution.

**Reply by maciejpiotrowski89** [reply]
@vfn I think the scenario you wrote about is called **Repository Grafting**. It was mentioned in the GitHub Universe workshops I linked in my initial message - here you can find a [list of steps](https://github.com/githubuniverseworkshops/grafting-monorepos#seedling-activity-3-graft-a-repository-20-minutes) required to perform it. The steps include "repository cleanup" - if you want to have a lightweight repository after the graft, you would have to e.g. track heavy BLOBs in **Git LFS**, as you already have mentioned. 

One of the cons of **Repo Grafting** is lack of old history when you clone the new repo, that contains only the newest history. The old history might be crucial for some projects or modules within a huge codebase. One can have both histories linked with some git magic though. 

I haven't found any information about grafting or even working with huge git repositories for mobile apps. Maybe other folks here can share their experience 🙂

**Reply by vfn** [reply]
@maciejpiotrowski89 I see what you mean, thanks for the link.

I was sort of familiar with the concept of [grafting in Git](https://git.wiki.kernel.org/index.php/GraftPoint), but never thought of it for the repo migration strategy I had in mind, and in fact it would fit perfectly to stitch the history together as/when needed.

> One of the cons of **Repo Grafting** is lack of old history when you clone the new repo, that contains only the newest history. The old history might be crucial for some projects or modules within a huge codebase. One can have both histories linked with some git magic though.

That's true, and it could be easily mitigated by explaining in the repo's README how to fetch the history that precedes the latest repo's one, and by adding a script to help with that task. As in the latest repo's script would know how to stitch back all the previous repos. When preparing the next repo, the team in charge would have as one of the tasks to add a reference to the previously latest repo to the script.

That could be a pain if most of the team needs access to the older history and if that was a common task, however, for people that don't usually need to look that far back in the repo's history, that would be a great win.

**Comment by brianduff** [comment]
Just thought I'd contribute a few more datapoints. 

At Twitter, we have a monorepo for service code, and three separate monorepos for iOS, Android, and Web.

Of places I've worked in the past:

Facebook uses a single monorepo for all service development, Android and iOS. Because of the way Buck works, these three trees in the monorepo have slightly different configuration, but dependencies across them are allowed with limitations. Web was still separate.
 
Google, of course, has for many years used a single monorepo with the exception of large open source things like Android's core (yet, almost all of Google's Android apps are in the monorepo) and Chrome. Google's monorepo is the most "integrated"  - e.g. for JVM code, it's structured as a unified single code tree that includes backend Java and Android code side-by-side.

Oracle didn't have a monorepo when I was there, which was a pretty long time ago, and I don't imagine that it does now (it also didn't have much in the way of mobile at least back then).

**Comment by shahzadmajeed** [comment]
Hi,

I really love this thread. Very informative!

I work at DraftKings and experimenting with mono-repo that will host several iOS applications (there will be a separate repo for Android platform). We are in bitbucket and planning to use CODEOWNERS so that might help with PR review process. We are not using bazel but in the process of setting up Tuist for some of the projects.

I have few questions below if you guys don't mind sharing experience/knowledge/wisdom. 

1. How does the release management looks like for multiple apps in same repo? Do they all release at same time or do you maintain difference release cycles for multiple apps? Is there a well-defined version management process for such a case?
2. Related to # 1 but how do you manage releases if only a specific app need to adopt changes in common code and other apps will still need to use older code?
3. How do you manage CI/CD for changes in common vs. app specific code?
4. How do you achieve caching in CI/CD?
5. How do you run code smell (linting, formatting) tools on mono-repo? Is it per app or for the whole repo? More specifically, linting whole repo when building app_1 could be unproductive and confusing to see errors from unrelated code. So maybe use child configurations for code smell tools?
7. Are there any tools that make managing mono-repos easy?
8. Anything else related to mono-repos?

 Thank you
-Shahzad

**Reply by keith** [reply]
Some quick answers for Lyft:

1. We create release branches per app still and release at whatever pace for each app from those. for us all app's release cycles are the same, but theoretically since they have their own branches it could be different
2. When we have release fixes we first merge into the main branch and then with the separate branches you have to intentionally cherry pick common code to whichever apps you want the fixes to apply to. This means more cherry pick PR work but we automated that away
3. We use some build system logic similar to https://github.com/Tinder/bazel-diff/ to get the list of changed targets based on what files were changed on a PR to always build the minimum set of things that could break given the PR
4. For us this is solved by the build system (bazel)
5. We run only on the git-diff in most cases locally to run faster and only spot things you've changed. but we still run everything in the case you change the rules or something themselves. This means it blocks PRs globally, and if that's a major migration pain we add opt-outs for the previous violations, but at least no new violations can make it in
6 & 7. Guess it depends on what you're worried about specifically

hope this helps!

**Reply by shahzadmajeed** [reply]
Thank you for taking time to reply. I have a follow up questions on # 2:

Cherry picking approach is interesting but how does that work if you have to apply a hotfix_2 for app_2 on top of hotfix_1 for app_1 but second hotfix_2 for app_2 still doesn't want hotfix_1. There could be some overlapping is the automation intelligent enough to raise such concerns? Is this tool open source or could you please provide some further pointers if not open source?

**Reply by keith** [reply]
Are you're saying that you need to combine 2 hotfixes on 1 app and not the other, but the 2 hotfixes are incompatible? If so in our case we definitely leave that up to the release manager to decide what of these should be merged etc, so they should be aware of those functional conflicts. But also theoretically both hotfix 1 and 2 have already landed for all apps on the main branch, so if they are incompatible you already have a problem for future releases

**Reply by shahzadmajeed** [reply]
Yes, incompatible hotfixes that are part of main branch but one of the apps doesn't want to include in future release.  
Sounds like it is some manual work for release manager to decide what to include in each app in this case.

**Reply by keith** [reply]
Realistically having a change that "one app doesn't want" is just not acceptable with this solution really. You just have to make sure your solution works for all cases. So really we don't hit collisions like this ever.
